webrtc/RTCTransport-helper.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+/*
+  Code using this helper should also include RTCPeerConnection-helper.js
+  in the main HTML file
+
+  The following helper functions are called from RTCPeerConnection-helper.js:
+    getTrackFromUserMedia
+    addTrackOrTransceiver
+    exchangeIceCandidates
+    doSignalingHandshake
+*/
+
+// Wait for all DTLS and ICE transport of a peer connection
+// to turn to connected or completed state. When the RTCPeerConnectionState
+// is "connected", all RTCIceTransports and RTCDtlsTransports are in the
+// "connected", "completed" or "closed" state and at least one of them is in the
+// "connected" or "completed" state.
+async function waitForConnectedState(pc) {
+  if (pc.connectionState === 'connected') {
+    return;
+  }
+
+  return new Promise((resolve, reject) => {
+    pc.addEventListener('connectionstatechange', () => {
+      const { connectionState } = pc;
+
+      if (connectionState === 'connected') {
+        resolve();
+      } else if (['closed', 'failed'].includes(connectionState)) {
+        reject(new Error(`one of DTLS/ICE transports transition to unexpected state ${connectionState}`));
+      }
+    });
+  });
+}
+
+function getDtlsTransportFromSctpTransport(sctpTransport) {
+  assert_true(sctpTransport instanceof RTCSctpTransport,
+    'Expect pc.sctp to be instantiated from RTCSctpTransport');
+
+  const dtlsTransport = sctpTransport.transport;
+  assert_true(dtlsTransport instanceof RTCDtlsTransport,
+    'Expect sctp.transport to be an RTCDtlsTransport');
+
+  return dtlsTransport;
+}
+
+function getIceTransportFromSctpTransport(sctpTransport) {
+  const dtlsTransport = getDtlsTransportFromSctpTransport(sctpTransport);
+
+  const { iceTransport } = dtlsTransport;
+  assert_true(iceTransport instanceof RTCIceTransport,
+    'Expect dtlsTransport.iceTransport to be an RTCIceTransport');
+
+  return iceTransport;
+}
+
+// Get one or more RTCDtlsTransports from RTCRtpSender/Receiver
+// in pc.There may be more than one DTLSTransport if there are
+// multiple underlying transports, such as when not using max-bundle,
+// or when not using RTCP multiplexing.
+function getDtlsTransportsFromSenderReceiver(pc) {
```

Comment:
These helpers collectively abstract away the peer connection APIs we're testing. That may be the right approach for application development, but for writing API tests, I prefer repetition operating on the APIs we're actually testing.

This applies to the remaining helpers here as well.

My 2 cents.

Reply:
I think it depends on how complicated it is to figure how to perform an operation correctly. There are many things in WebRTC that are not as straightforward as "I want to do X, therefore I'd just call method Y".

I have removed the helpers and make them inline. The inline transport extractors are simplified and not doing any thorough checking or assertions anymore. Hope this will be more readable, easier to verify correctness, and easy for copy/paste for new test cases.